### PR TITLE
Remove instantiable Address

### PR DESCRIPTION
**Pull request by dcousens**
~~WIP pull request.~~ Ready to merge

### CODE REVIEW

File: src/address.js [hunk 1]
```diff
@@ -1,62 +1,67 @@
...
-  var scriptType = findScriptTypeByVersion(this.version)
+    var version = payload.readUInt8(0)
 
-  if (scriptType === 'pubkeyhash') return scripts.pubKeyHashOutput(this.hash)
-  if (scriptType === 'scripthash') return scripts.scriptHashOutput(this.hash)
+    assert(version === network.pubKeyHash || version === network.scriptHash)
 
-  assert(false, this.toString() + ' has no matching Script')
+  } catch (e) {
+    throw new Error(address + ' is not a valid ' + network + ' address')
```
Comment (weilu) [inline src/address.js]: Does validate have to throw on invalid? Consider returning a boolean instead?

File: src/address.js [hunk 2]
```diff
@@ -1,62 +1,67 @@
...
-function Address (hash, version) {
-  typeForce('Buffer', hash)
+  } else if (scripts.isScriptHashOutput(script)) {
+    version = network.scriptHash
+    hash = script.chunks[1]
 
-  assert.strictEqual(hash.length, 20, 'Invalid hash length')
-  assert.strictEqual(version & 0xff, version, 'Invalid version byte')
+  } else {
+    assert(false, script.toASM() + ' has no matching Address')
```
Comment (weilu) [inline src/address.js]: s/Address/address

File: src/address.js (hunk 1 above)
Comment (dcousens) [inline src/address.js]: Sure.

**Comment by weilu** [comment]
@dcousens Let's finish this.

**Comment by dcousens** [comment]
@weilu sounds good,  I'll get back on this today.

**Comment by dcousens** [comment]
Tests failing for unrelated reasons.  I intend to fix that.

@weilu please review.
Should be quickest to review commit-by-commit.

The API could maybe change,  but lets do that in a different PR,  I want to get this merged and leave the API bike shedding separate. 

@jprichardson feel free to review also. :)

**Comment by dcousens** [comment]
Intend to merge ASAP.  Any comments/qualms?

**Comment by jprichardson** [comment]
Most of it looks good. I like this change. My only thought is not removing the tests (you'd still keep this commit as is https://github.com/bitcoinjs/bitcoinjs-lib/commit/16a711ba3309e0a4b8f0751cab5b1ab6e7366c40 i.e. remove these tests as you have). But by not removing the other tests and maybe just changing them to errors helps to communicate how the expectations have changed. I'm fine with it as is, just a thought.

Also, what's with bundling the commit to add the network into TxBuilder into this PR?

**Comment by dcousens** [comment]
@jprichardson please see above,  I rebased the `TransactionBuilder` commit so it is inline with the other changes which I hadn't pushed! (Good spot! :))

**Comment by dcousens** [comment]
@jprichardson never mind,  I rebased without that commit and will submit in a different PR.

File: src/address.js [hunk 3]
```diff
@@ -13,50 +12,55 @@ function findScriptTypeByVersion (version) {
...
-Address.fromBase58Check = function (string) {
+function fromBase58Check (string) {
   var payload = base58check.decode(string)
+  if (payload.length !== 21) throw new TypeError('Invalid address length')
+
   var version = payload.readUInt8(0)
   var hash = payload.slice(1)
 
-  return new Address(hash, version)
+  return { hash: hash, version: version }
```
Comment (weilu) [inline src/address.js]: Not a big fan of the return value of this function. How about splitting this into two functions: `hashFromBase58Check` and `versionFromBase58Check`

File: src/address.js [hunk 4]
```diff
@@ -13,50 +12,55 @@ function findScriptTypeByVersion (version) {
...
 }
 
-Address.fromOutputScript = function (script, network) {
+function fromOutputScript (script, network) {
   network = network || networks.bitcoin
 
-  if (scripts.isPubKeyHashOutput(script)) return new Address(script.chunks[2], network.pubKeyHash)
-  if (scripts.isScriptHashOutput(script)) return new Address(script.chunks[1], network.scriptHash)
+  if (scripts.isPubKeyHashOutput(script)) return toBase58Check(script.chunks[2], network.pubKeyHash)
+  if (scripts.isScriptHashOutput(script)) return toBase58Check(script.chunks[1], network.scriptHash)
```
Comment (weilu) [inline src/address.js]: Similarly here, an object with `hash` and `version` seems like a half backed internal representation. I'd rather this returning a base58 encoded string instead.

**Comment by weilu** [comment]
There will be a PR for adding deprecation warnings on 1.x branch right?

**Comment by dcousens** [comment]
`fromBase58Check` is temporary IMHO.

> There will be a PR for adding deprecation warnings on 1.x branch right?

Of course. 

> Add a possibleNetworks here as the 2nd argument, and pass it to findNetworkByWIFVersion for network detection.

I think that's worth investigating,  but we need to do *consistently* this across all the network sections, and I'd be happy to do that in https://github.com/bitcoinjs/bitcoinjs-lib/pull/425.

File: src/address.js (hunk 4 above)
Comment (dcousens) [inline src/address.js]: Huh?  It doesn't return an `object`,  it returns a base58check encoded string.

**Comment by dcousens** [comment]
Going to merge, as the issues brought up (here and privately) were all expected:

* `fromBase58check` feels like an internal API,  but it still useful,  lets find a decent replacement (PR to come).
* Network detection cleanup (will be addressed in https://github.com/bitcoinjs/bitcoinjs-lib/pull/425)
* Deprecation warnings (will put a note on the `1.x` branch for pre-2.0.0 tasks)

**Comment by dcousens** [comment]
@coveralls, you trippin

**Pull request merged**